src/ctypes/ldouble_stubs.c

```diff
@@ -0,0 +1,525 @@
+  case FP_INFINITE  : r = Val_int(ml_FP_INFINITE); break;
+  case FP_NAN       : 
+  default           : r = Val_int(ml_FP_NAN); break;
+  }
+  CAMLreturn(r);
+}
+
+static char *format_ldouble(char *fmt, long double d) {
+  static size_t buf_len = 10;
+  static char *buf = NULL;
+  static char *empty = "";
+  size_t print_len = 0;
+
+  // allocate initial string
+  if (buf == NULL) buf = malloc(buf_len);
+  if (buf == NULL) return empty; // oops...stuck
+
+  // try to print
+  print_len = snprintf(buf, buf_len, fmt, d);
+
+  // not enough space - reallocate and try again
+  if (print_len >= buf_len) {
+    // re-allocate string
+    if (buf) free(buf);
+    buf = malloc(print_len+1);
+    buf_len = print_len+1;
+    // try again
+    return format_ldouble(fmt, d);
+  } else 
+    // ok
+    return buf;
+}
+
+CAMLprim value ctypes_ldouble_format(value fmt, value d) {
+  CAMLparam2(fmt, d);
+  CAMLreturn(caml_copy_string(format_ldouble( String_val(fmt), ldouble_custom_val(d))));
+}
+
+CAMLprim value ctypes_ldouble_of_string(value v) {
+  CAMLparam1(v);
+  char *str = String_val(v);
+  char *end = str + caml_string_length(v);
+  long double r = strtold(str, &end);
+  CAMLreturn(ctypes_copy_ldouble(r));
+}
+
+char hex_char(char x) {
+  if (x < 10) return '0' + x;
+  return 'a' + x - 10;
+}
+
+CAMLprim value ctypes_ldouble_to_hex(value v) {
+  CAMLparam1(v);
+  static char x[LDOUBLE_STORAGE_BYTES*2 + 1];
+  char *p = (char *) Data_custom_val(v);
+  int i;
+  for (i=0; i<LDOUBLE_STORAGE_BYTES; i++) {
+    x[i*2+0] = hex_char(((*(p+i)) >> 0) & 0xf);
+    x[i*2+1] = hex_char(((*(p+i)) >> 4) & 0xf);
+  }
+  x[LDOUBLE_STORAGE_BYTES*2] = 0;
+  CAMLreturn(caml_copy_string(x));
+}
+
+value ctypes_ldouble_min(value unit) { return ctypes_copy_ldouble(-LDBL_MAX); }
+value ctypes_ldouble_max(value unit) { return ctypes_copy_ldouble(LDBL_MAX); }
+value ctypes_ldouble_epsilon(value unit) { return ctypes_copy_ldouble(LDBL_EPSILON); }
+value ctypes_ldouble_nan(value unit) { return ctypes_copy_ldouble(nanl("char-sequence")); }
+// XXX note; -(log 0) gives +ve inf (and vice versa).  Is this consistent? *)
+value ctypes_ldouble_inf(value unit) { return ctypes_copy_ldouble(-log(0)); } 
+value ctypes_ldouble_ninf(value unit) { return ctypes_copy_ldouble(log(0)); }
+
+value ctypes_ldouble_size(value unit) {
+  CAMLparam1(unit);
+  CAMLlocal1(r);
+  r = caml_alloc_tuple(2);
+  Store_field(r,0, Val_int(LDOUBLE_STORAGE_BYTES));
+  Store_field(r,1, Val_int(LDOUBLE_VALUE_BYTES));
+  CAMLreturn(r);
+}
+
+/*********************** complex *************************/
+
+#define ldouble_complex_custom_val(V) (*(long double complex*)(Data_custom_val(V)))
+
+static int ldouble_complex_cmp_val(value v1, value v2)
+{
+  long double complex u1 = ldouble_custom_val(v1);
+  long double complex u2 = ldouble_custom_val(v2);
+  int cmp_real = ldouble_cmp(creall(u1), creall(u2));
+  return cmp_real == 0 ? ldouble_cmp(cimagl(u1), cimagl(u2)) : cmp_real;
```

Comment:
This doesn't follow the OCaml conventions for NaNs (but fixing the `ldouble` comparison should fix this one without additional changes).